File: backend/src/main/java/com/smartcampus/service/BookingService.java
package com.smartcampus.service;

import com.smartcampus.dto.BookingResponse;
import com.smartcampus.dto.CreateBookingRequest;

import java.time.LocalDate;
import java.util.List;

public interface BookingService {
    BookingResponse createBooking(long userId, CreateBookingRequest request);
    BookingResponse cancelBooking(long bookingId, long userId);
    BookingResponse approveBooking(long bookingId, String adminNote);
    BookingResponse rejectBooking(long bookingId, String adminNote);
    BookingResponse adminCancelBooking(long bookingId);
    List<BookingResponse> getMyBookings(long userId);
    List<BookingResponse> getBookingsByResourceAndDate(String resourceId, LocalDate date);
    List<BookingResponse> getAllBookings();
    BookingResponse getBookingById(long id, long userId, boolean isAdmin);
    void deleteBooking(long bookingId, long userId, boolean isAdmin);
}

File: frontend/src/services/bookingService.js
import api from './api';

const getAuthHeaders = () => {
  const currentUserRaw = localStorage.getItem('current_user');

  if (!currentUserRaw) {
    return {};
  }

  try {
    const currentUser = JSON.parse(currentUserRaw);
    return {
      'X-User-Id': currentUser.id,
      'X-User-Role': currentUser.role,
    };
  } catch (error) {
    return {};
  }
};

const authConfig = (extraConfig = {}) => ({
  ...extraConfig,
  headers: {
    ...(extraConfig.headers || {}),
    ...getAuthHeaders(),
  },
});

export const bookingAPI = {
  createBooking: (data) => api.post('/bookings', data, authConfig()),
  getMyBookings: () => api.get('/bookings/my', authConfig()),
  getBookingById: (id) => api.get(`/bookings/${id}`, authConfig()),
  cancelBooking: (id) => api.patch(`/bookings/${id}/cancel`, {}, authConfig()),
  deleteBooking: (id) => api.delete(`/bookings/${id}`, authConfig()),
  getAllBookings: () => api.get('/bookings', authConfig()),
  approveBooking: (id, adminNote) => api.patch(`/bookings/${id}/approve`, { adminNote }, authConfig()),
  rejectBooking: (id, adminNote) => api.patch(`/bookings/${id}/reject`, { adminNote }, authConfig()),
  adminCancelBooking: (id) => api.patch(`/bookings/${id}/admin-cancel`, {}, authConfig()),
  getResourceBookings: (resourceId, date) => api.get(`/bookings/resource/${resourceId}`, {
    params: { date },
    ...authConfig(),
  }),
};
